src/ratelimit_manager.cpp

```diff
@@ -0,0 +1,388 @@
+            }
+ 
+            // Remove ban
+            throttled_entities.erase(entity);
+
+
+            // Reset request counts
+            auto& request_counts = rate_limit_request_counts.lookup(entity);
+            request_counts.reset();
+
+
+            // Check if throttle rule still exists
+            if (rule.action == RateLimitAction::throttle) {
+                // Increase because of this request
+                request_counts.current_requests_count_minute++;
+                request_counts.current_requests_count_hour++;
+            }
+
+            continue;
+
+        }
+        else {
+            
+            if(!rate_limit_request_counts.contains(entity)){
+                rate_limit_request_counts.insert(entity, request_counter_t{});
+            }
+            
+            auto& request_counts = rate_limit_request_counts.lookup(entity);
+
+            // Check if last reset time was more than 1 minute ago
+            if (request_counts.last_reset_time_minute <= std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 60) {
+                request_counts.previous_requests_count_minute = request_counts.current_requests_count_minute;
+                request_counts.current_requests_count_minute = 0;
+
+                if(request_counts.last_reset_time_minute < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 120) {
+                    request_counts.previous_requests_count_minute = 0;
+                }
+
+                request_counts.last_reset_time_minute = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+            }
+
+            // Check if last reset time was more than 1 hour ago
+            if (request_counts.last_reset_time_hour < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 3600) {
+                request_counts.previous_requests_count_hour = request_counts.current_requests_count_hour;
+                request_counts.current_requests_count_hour = 0;
+
+                if(request_counts.last_reset_time_hour < std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - 7200) {
+                    request_counts.previous_requests_count_hour = 0;
+                }
+
+                request_counts.last_reset_time_hour = std::chrono::system_clock::to_time_t(std::chrono::system_clock::now());
+            }
+
+
+
+            // Check if request count is over the limit
+            auto current_rate_for_minute = (60 - (std::chrono::system_clock::to_time_t(std::chrono::system_clock::now()) - request_counts.last_reset_time_minute)) / 60  * request_counts.previous_requests_count_minute;
+            current_rate_for_minute += request_counts.current_requests_count_minute;
+
+            if(rule.max_requests.minute_threshold >= 0 && current_rate_for_minute >= rule.max_requests.minute_threshold) {
+                if(rule.auto_ban_threshold_num >= 0 && rule.auto_ban_num_days >= 0) {
```

Comment:
`auto_ban_threshold_num` must actually be greater than 0, because `auto_ban_threshold_num: 0` has no meaning. Likewise for `auto_ban_num_days`.

Also, can we extract this into a meaningful boolean variable like this?

```
bool auto_ban_is_enabled = (rule.auto_ban_threshold_num > 0 && rule.auto_ban_num_days > 0);
```

Helps with readability.